cpp/src/arrow/compute/kernels/row_encoder.cc

```diff
@@ -255,6 +256,166 @@ Result<std::shared_ptr<ArrayData>> DictionaryKeyEncoder::Decode(uint8_t** encode
+      const auto& scalar = data.scalar_as<ListScalar>();
+      std::shared_ptr<ArrayData> child_data = scalar.value->data();
+      list_length = child_data->length;
+      int64_t child_size = child_data->type->byte_width();
+      int64_t buffer_size = list_length * (child_size + 1);
+
+      val.reserve(buffer_size);
+      for (int64_t i = 0; i < list_length; i++) {
+        if (child_data->IsValid(i)) {
+          val.push_back(kValidByte);
+          for (int64_t k = 0; k < child_size; k++) {
+            val.push_back(static_cast<char>(child_data->buffers[1]->data()[i * child_size + k]));
+          }
+        }
+        else {
+          val.push_back(kNullByte);
+          for (int64_t k = 0; k < child_size; k++) {
+            val.push_back(static_cast<char>(0));
+          }            
+        }
+      }
+    }
+
+    for (int64_t i = 0; i < batch_length; i++) {
+      auto& encoded_ptr = *encoded_bytes++;
+      *encoded_ptr++ = data.scalar->is_valid ? kValidByte : kNullByte;
+      util::SafeStore(encoded_ptr, static_cast<Offset>(list_length));
+      encoded_ptr += sizeof(Offset);
+      memcpy(encoded_ptr, &val[0], val.size());
+      encoded_ptr += static_cast<int64_t>(val.size());
+    }
+  }
+  return Status::OK();
+}
+
+void ListFixedLengthChildEncoder::EncodeNull(uint8_t** encoded_bytes) {
+  auto& encoded_ptr = *encoded_bytes;
+  *encoded_ptr++ = kNullByte;
+  util::SafeStore(encoded_ptr, static_cast<Offset>(0));
+  encoded_ptr += sizeof(Offset);
+}
+
+Result<std::shared_ptr<ArrayData>> ListFixedLengthChildEncoder::Decode(uint8_t** encoded_bytes, int32_t length,
+                                          MemoryPool* pool) {
+  std::shared_ptr<DataType> child_type = std::dynamic_pointer_cast<ListType>(type_)->value_type();
+  Offset child_size = child_type->byte_width();
+
+  std::shared_ptr<Buffer> null_buf;
+  int32_t null_count;
+  ARROW_RETURN_NOT_OK(DecodeNulls(pool, length, encoded_bytes, &null_buf, &null_count));
+
+  Offset length_sum = 0;
+  for (int64_t i = 0; i < length; ++i) {
+    length_sum += util::SafeLoadAs<Offset>(encoded_bytes[i]);
+  }
+
+  ARROW_ASSIGN_OR_RAISE(auto offset_buf,
+                        AllocateBuffer(sizeof(Offset) * (1 + length), pool));
+  ARROW_ASSIGN_OR_RAISE(auto key_buf, AllocateBuffer(length_sum * child_size, pool));
+  ARROW_ASSIGN_OR_RAISE(auto child_null_buf, AllocateBitmap(length_sum, pool));
+
+  auto raw_offsets = reinterpret_cast<Offset*>(offset_buf->mutable_data());
+  auto raw_keys = key_buf->mutable_data();
+  auto raw_child_validity = child_null_buf->mutable_data();
+
+  FirstTimeBitmapWriter writer(raw_child_validity, 0, length_sum);
+  Offset current_child_offset = 0;
+  int64_t child_null_count = 0;
+  for (int64_t i = 0; i < length; ++i) {
+    raw_offsets[i] = current_child_offset;
+
+    auto list_length = util::SafeLoadAs<Offset>(encoded_bytes[i]);
+    encoded_bytes[i] += sizeof(Offset);
+
+    for (int64_t j = 0; j < list_length; j++) {
```

Comment:
after reading this code, I wonder if a better encoding scheme would be to encode the entire null bitmap first, then the child values array...

you might be able to make this code a bit simpler to read by directly increment `current_child_offset` in the loop instead of keeping track of a separate loop counter, but I don't feel too strongly

Reply:
I agree about encoding the entire null bitmap first and will do. I'll have to be able to write my own memcpy for bits.

Reply:
I thought about the encoding scheme about encoding the entire null bitmap first and here are my thoughts:

1. Storing the null bitmap first is more memory efficient (storing a bit per element instead of a byte).
2. Will need to implement a memcpy for bits (in bits_util).
3. Not having null bitmap before all the elements is more iterative. (When decoding can have 1 pointer instead of a pointer pointing at the null bitmap and at the elements)
4. I imagine for list<string> it be null bit map, then string lengths and then elements

And for `current_child_offset`, I wanted the value to store at the start of every list (the child being list), so I didn't want to increment it. (The name may be confusing and may be changed to `current_list_offset`)

Reply:
For (2) can't we just memcpy from the encoded bytes buffer to `child_null_buf` instead of going through bits_util? I was imagining that the encoding would literally just be the nullity buffer concatenated with the data buffer. And I was hoping this would simplify the code if you could encode/decode elements via two memcpy calls instead of doing it iteratively.

Ultimately though, I think the current encoding scheme does work. So I'm okay with just going with this and revisiting later if we want to optimize.

Reply:
ok ignore me I just realized why this wouldn't work: the null bits associated with a single list may not be byte aligned. so those bits would still need to be manually encoded/decoded (though this still doesn't seem so bad and the data buffer can be memcopied).